### PR TITLE
streaming summarizer performance

### PR DESCRIPTION
**Pull request by edgan8**
Added a unit test for comparing streaming perf with batch perf
default values tuned for fast default unit test time, increase
to evaluate more realistic workloads.

Also fixed bug with dataframe union that was leading to issues with
windowing operator implementation

### CODE REVIEW

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/itemset/result/AttributeSet.java [hunk 1]
```diff
@@ -72,4 +72,11 @@ public String toString() {
                 ", items=" + items +
                 '}';
     }
+
+    @Override
+    public int compareTo(AttributeSet o) {
```
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/itemset/result/AttributeSet.java]: going to implement sorting these explanation itemsets soon

File: lib/src/main/java/edu/stanford/futuredata/macrobase/datamodel/DataFrame.java [hunk 2]
```diff
@@ -169,35 +169,35 @@ public static DataFrame unionAll(List<DataFrame> others) {
...
         combined.numRows = n;
+        int d = first.schema.getNumColumns();
 
-        int d = first.stringCols.size();
-        combined.stringCols = new ArrayList<>(d);
         for (int colIdx = 0; colIdx < d; colIdx++) {
-            String[] newCol = new String[n];
-            int i = 0;
-            for (DataFrame curOther : others) {
-                String[] otherCol = curOther.getStringColumn(colIdx);
```
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/datamodel/DataFrame.java]: the bug here was that you need to index columns by their global index, not by the internal index when using these apis.

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/IncrementalSummarizer.java [hunk 3]
```diff
@@ -41,26 +41,35 @@
...
 
+    public IncrementalSummarizer() {
+        setWindowSize(1);
+    }
+
+
     @Override
     public void setWindowSize(int numPanes) {
         this.numPanes = numPanes;
+        initializePanes();
```
Comment (kexinrong) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/IncrementalSummarizer.java]: We only want to call this function if `inlierPaneCounts, outlierPaneCounts` etc. are not initialized.

File: lib/src/test/java/edu/stanford/futuredata/macrobase/StreamingSummarizationTest.java [hunk 4]
```diff
@@ -80,6 +91,74 @@ public static DataFrame generateAnomalyDataset(
...
+            timerStart = System.currentTimeMillis();
+            bsumm.process(curWindow);
+            Explanation batchExplanation = bsumm.getResults();
+            timerElapsed = System.currentTimeMillis() - timerStart;
+            totalBatchTime += timerElapsed;
+
+            startTime = endTime;
+        }
+
+        assertTrue(totalStreamingTime < totalBatchTime);
```
Comment (kexinrong) [inline lib/src/test/java/edu/stanford/futuredata/macrobase/StreamingSummarizationTest.java]: This assertion is failing somehow?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java [hunk 5]
```diff
@@ -19,12 +21,46 @@ public Explanation(List<AttributeSet> resultList,
...
+     * Removes redundant explanations
+     * @return New explanation with redundant itemsets removed.
+     */
+    public Explanation prune() {
+        List<AttributeSet> newItemsets = new ArrayList<>();
+        int n = itemsets.size();
+        for (int i = 0; i < n; i++) {
+            AttributeSet aSet = itemsets.get(i);
+            boolean redundant = false;
+            // an explanation is redundant if it has lower risk ratio (occurs after since sorted)
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java]: are you sure itemsets are sorted here?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/datamodel/DataFrame.java (hunk 2 above)
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/datamodel/DataFrame.java]: good catch!

File: lib/src/main/java/edu/stanford/futuredata/macrobase/operator/WindowedOperator.java [hunk 6]
```diff
@@ -137,5 +137,12 @@ public double getMaxWindowTime() {
     public int getBufferSize() {
         return batchBuffer.size();
     }
+    public int getBufferedRows() {
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/operator/WindowedOperator.java]: getNumBufferedRows?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java (hunk 5 above)
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java]: Yeah, I figured for convenience we can sort the explanation in the constructor and always have it sorted.

Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java]: as long as no one rearranges it it :)

**Comment by pbailis** [comment]
Good to go @edgan8

**Pull request merged**